### PR TITLE
Retry hash file allocation

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

Recently, on Oct 6, 20023, we saw a validator crash on one of the canaries boxes. 

```
[2023-10-06T06:29:48.605280562Z INFO  solana_metrics::metrics] datapoint: accounts_db_active shrink=0i
[2023-10-06T06:29:48.653273411Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="solAccountsLo03" one=1i message="panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 12, kind: OutOfMemory, message: \"Cannot allocate memory\" }', accounts-db/src/accounts_hash.rs:102:34" location="accounts-db/src/accounts_hash.rs:102:34" version="1.18.0 (src:e0091d69; feat:1091887072, client:SolanaLabs)"
```

The crash happens when allocating account hash file during hash dedup. 

```
[2023-10-06T06:29:31.721389079Z INFO  solana_metrics::metrics] datapoint: memory-stats total=269907476480i swap_total=0i free_percent=3.6710358146504354 used_bytes=39749242880i avail_percent=85.27301155255498 buffers_percent=0.6657022085615106 cached_percent=79.78760422961368 swap_free_percent=0
[2023-10-06T06:29:36.744595335Z INFO  solana_metrics::metrics] datapoint: memory-stats total=269907476480i swap_total=0i free_percent=5.081942011716148 used_bytes=35619844096i avail_percent=86.80294278597377 buffers_percent=0.6663881443585271 cached_percent=79.90525739437273 swap_free_percent=0
[2023-10-06T06:29:41.745857461Z INFO  solana_metrics::metrics] datapoint: memory-stats total=269907476480i swap_total=0i free_percent=5.645556638416836 used_bytes=35985874944i avail_percent=86.66732933325522 buffers_percent=0.6665914970063151 cached_percent=79.21095985788381 swap_free_percent=0
[2023-10-06T06:29:46.768160791Z INFO  solana_metrics::metrics] datapoint: memory-stats total=269907476480i swap_total=0i free_percent=0.6511670469158839 used_bytes=37269983232i avail_percent=86.19157063818434 buffers_percent=0.6666142603624108 cached_percent=83.64288057827423 swap_free_percent=0
```

The `used` memory increase by 2G and the free memory drops to 0.65 percent. 

It appears that the box still has enough physical memory. However, the allocation fails. Probably due to memory defragmentation and multiple threads allocate mmaps at the same time. 

#### Summary of Changes

Account hash calculation is highly parallel. It seems that when all threads start allocation files to store account hashes at the same time, the system may hit OOM for some of those allocations. To mitigate that, we add retry for account hash file allocation. If a thread fails its allocation, it will sleep for some time and then retry. This will help to stagger the memory allocation from all parallel threads.

Hopefully, the kernel has time to defrag the memory in-between the retries, and the later retry for allocation will succeed. 

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_hash.rs [hunk 1]
```diff
@@ -87,21 +87,54 @@ impl AccountHashesFile {
...
-            data.rewind().unwrap();
-            data.flush().unwrap();
+            // Retry 5 times for allocation the AccountHashFile. The memory maybe fragmented and
+            // causes memory allocation failure. Therefore, retry after failure. Hoping that the
+            // kernel can defrag the memory and allocation retries can succeed.
+            let mut num_retries = 0;
+            let data = loop {
+                num_retries += 1;
+
+                match get_file() {
```
Comment (norwnd) [inline accounts-db/src/accounts_hash.rs:116]: Perhaps introducing some delay between retries is desirable (maybe even linearly increasing too) ?

Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs:117]: yeah. good idea.

**Comment by jeffwashington** [comment]
This seems reasonable to me. I'd like to see a data point published if we actually have to retry.

File: accounts-db/src/accounts_hash.rs [hunk 2]
```diff
@@ -87,22 +88,56 @@ impl AccountHashesFile {
...
-            // Theoretical performance optimization: write a zero to the end of
-            // the file so that we won't have to resize it later, which may be
-            // expensive.
-            assert!(self.capacity > 0);
-            data.seek(SeekFrom::Start((self.capacity - 1) as u64))
-                .unwrap();
-            data.write_all(&[0]).unwrap();
-            data.rewind().unwrap();
-            data.flush().unwrap();
+            // Retry 5 times for allocation the AccountHashFile. The memory maybe fragmented and
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:110]: typo

File: accounts-db/src/accounts_hash.rs [hunk 3]
```diff
@@ -87,22 +88,56 @@ impl AccountHashesFile {
...
+                        );
+
+                        if num_retries > 5 {
+                            panic!(
+                                "Unable to create account hash file within {}: after {} retries",
+                                self.dir_for_temp_cache_files.display(),
+                                num_retries
+                            );
+                        }
+                        datapoint_info!("account_hash_file_allocation_retry", ("retry", 1, i64),);
```
Comment (jeffwashington) [inline accounts-db/src/accounts_hash.rs:136]: trailing , to remove since there's another typo above.

File: accounts-db/src/accounts_hash.rs (hunk 2 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_hash.rs]: Original:
```
Retry 5 times for allocation the..
```
suggestion:
Retry 5 times to allocate the...

Maybe more grammar than typo.

Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs]: ok. done.

File: accounts-db/src/accounts_hash.rs [hunk 4]
```diff
@@ -87,22 +88,56 @@ impl AccountHashesFile {
...
+
+                        if num_retries > 5 {
+                            panic!(
+                                "Unable to create account hashes file within {}: after {} retries",
+                                self.dir_for_temp_cache_files.display(),
+                                num_retries
+                            );
+                        }
+                        datapoint_info!("retry_account_hashes_file_allocation", ("retry", 1, i64));
+                        thread::sleep(time::Duration::from_millis(num_retries * 100));
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_hash.rs:137]: yeah. I think this should not happen very often, and choosing 100ms to be conservative.

**Comment by HaoranYi** [comment]
It seems that a similar crash happened again on mc3 last Friday. 
@jeffwashington @brooksprumo Do you mind if we merge this PR and give it a try on the canary box?

**Review by jeffwashington — APPROVED** [review]
lgtm

**Pull request merged**